File: sfm-splat-app/backend/core/viewpoint.py
"""viewpoint.py — the saved camera of the splat preview (CLAUDE.md §7.6d).

One thing, stored once, used in two places: the camera the user parked the
viewer at on step 4, and what the export writes beside — or inside — the file it
hands somebody else.

── Which frame it is stored in, which is the only subtle part ───────────────

The viewer draws everything under one `Rx-90` scene root plus the "Flip up"
toggle (`frontend/.../frame.ts`, §7.3), so a camera read straight off
`viewer.camera` is in **viewer** space and would mean a different thing the next
time the scene was opened with the flip the other way — exactly the trap the
crop volumes are stored in the dataset frame to avoid (§7.6b).

So a viewpoint is stored in the **dataset** frame: the +Z-up frame spirula's
mapper wrote, the frame `x, y, z` in the PLY are in, and the frame the crop
volumes already use. `viewpoint.ts` converts on both sides; nothing here has to
know about three.js.

── What "used on export" means, per format ─────────────────────────────────

* A native `.ply` carries it in its **header**, as `comment viewpoint …` lines.
  PLY comments are part of the format and every reader skips the ones it does
  not know, so this costs a few dozen bytes and breaks nothing.
* Every other format — `.splat`, and the three `splat-transform` writes — has
  nowhere to put it, so the export writes a `<name>.viewpoint.json` **beside**
  the file. A sidecar is honest about what it is; inventing a private container
  would not be.

Pure module: no FastAPI import (§2.4), and no numpy either — this is six floats
and a bool.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional, Sequence

#: The one place the sidecar's name is built. `.viewpoint.json` rather than
#: `.json`, so a drawer holding five exports never leaves anybody guessing which
#: file a bare `slug_sh0.json` belongs to.
SIDECAR_SUFFIX = ".viewpoint.json"

#: What the header comments and the sidecar both say about the numbers below.
FRAME_NOTE = "dataset frame (+Z up), the frame spirula wrote"


class ViewpointError(ValueError):
    """A stored viewpoint this module refuses to use."""


def _triple(raw: Any, name: str) -> tuple[float, float, float]:
    if not isinstance(raw, (list, tuple)) or len(raw) != 3:
        raise ViewpointError(f"viewpoint.{name}: expected three numbers")
    out = []
    for value in raw:
        if not isinstance(value, (int, float)) or isinstance(value, bool):
            raise ViewpointError(f"viewpoint.{name}: expected three numbers")
        value = float(value)
        if value != value or value in (float("inf"), float("-inf")):
            raise ViewpointError(f"viewpoint.{name}: not finite")
        out.append(value)
    return (out[0], out[1], out[2])


@dataclass(frozen=True)
class Viewpoint:
    """Where the user parked the camera, in the dataset frame."""

    position: tuple[float, float, float]
    target: tuple[float, float, float]
    #: The camera's up vector. Stored rather than assumed: the viewer's up is
    #: three.js's +Y, which is dataset +Z or -Z depending on "Flip up".
    up: tuple[float, float, float]
    #: Vertical field of view, degrees — three.js's `PerspectiveCamera.fov`.
    fov_y: float
    #: Which vertical was on when it was saved. Informative: the numbers above
    #: are already frame-independent, and this is what lets the viewer restore
    #: the toggle as well as the camera.
    flip_up: bool = False
    saved_at: Optional[str] = None

    def as_dict(self) -> dict:
        return {
            "position": list(self.position),
            "target": list(self.target),
            "up": list(self.up),
            "fov_y": self.fov_y,
            "flip_up": self.flip_up,
            "saved_at": self.saved_at,
        }

    # ── The two ways it reaches a file ──────────────────────────────────────

    def ply_comments(self) -> list[str]:
        """The `comment` lines a native PLY export carries in its header.

        Deliberately one key per line and space-separated: a PLY comment is
        free text, so the only thing that makes it readable by anything else is
        being trivially parseable — `comment viewpoint position X Y Z`.
        """
        def fmt(v: Sequence[float]) -> str:
            return " ".join(f"{c:.6g}" for c in v)

        lines = [
            f"viewpoint frame {FRAME_NOTE}",
            f"viewpoint position {fmt(self.position)}",
            f"viewpoint target {fmt(self.target)}",
            f"viewpoint up {fmt(self.up)}",
            f"viewpoint fov_y {self.fov_y:.6g}",
        ]
        if self.saved_at:
            lines.append(f"viewpoint saved_at {self.saved_at}")
        return lines

    def sidecar(self, target_file: Path) -> dict:
        return {
            "generated_at": datetime.now(timezone.utc).isoformat(),
            "file": target_file.name,
            "frame": FRAME_NOTE,
            **self.as_dict(),
        }


def parse(raw: Any) -> Optional[Viewpoint]:
    """A stored viewpoint, or None when there is none.

    Raises `ViewpointError` for a value that is there and unusable, because an
    export that silently drops the camera the user saved is the failure this
    whole feature exists to avoid.
    """
    if raw is None:
        return None
    if not isinstance(raw, dict):
        raise ViewpointError("viewpoint: expected an object")

    position = _triple(raw.get("position"), "position")
    target = _triple(raw.get("target"), "target")
    up = _triple(raw.get("up"), "up") if raw.get("up") is not None else (0.0, 0.0, 1.0)

    if position == target:
        raise ViewpointError("viewpoint: the camera is standing on its target")

    fov = raw.get("fov_y", 50.0)
    if not isinstance(fov, (int, float)) or isinstance(fov, bool):
        raise ViewpointError("viewpoint.fov_y: expected a number")
    fov = float(fov)
    if not 1.0 <= fov <= 179.0:
        raise ViewpointError(f"viewpoint.fov_y: {fov:g} is not a usable field of view")

    saved_at = raw.get("saved_at")
    return Viewpoint(
        position=position, target=target, up=up, fov_y=fov,
        flip_up=bool(raw.get("flip_up")),
        saved_at=saved_at if isinstance(saved_at, str) else None,
    )


def from_settings(settings: dict) -> Optional[Viewpoint]:
    """The viewpoint out of `settings_json`, or None.

    `viewpoint` is layer 3 only, with no `defaults.json` counterpart — for the
    crop volumes' reason (§7.6b): a camera parked in front of *this* scene is
    not a default anything could inherit.
    """
    raw = settings.get("viewpoint") if isinstance(settings, dict) else None
    return parse(raw)


def write_sidecar(view: Viewpoint, target_file: Path) -> Path:
    """Write `<name>.viewpoint.json` beside an exported file. Returns its path."""
    path = target_file.with_name(target_file.name + SIDECAR_SUFFIX)
    path.write_text(
        json.dumps(view.sidecar(target_file), indent=2), encoding="utf-8",
    )
    return path


def describe(view: Viewpoint) -> str:
    """One line for the log, in the dataset frame's own numbers."""
    pos = ", ".join(f"{c:.3g}" for c in view.position)
    tgt = ", ".join(f"{c:.3g}" for c in view.target)
    return f"({pos}) looking at ({tgt}), {view.fov_y:.0f}° vertical"

File: sfm-splat-app/frontend/src/components/viewer/viewpoint.ts
import * as THREE from 'three';
import { upRotationX } from './frame';

/**
 * The saved camera of the splat preview (CLAUDE.md §7.6d).
 *
 * One button on the viewer's toolbar, one object in `settings_json`, and one
 * thing the export writes into — or beside — the file it hands somebody else.
 *
 * ── The frame, which is the same care the crop volumes needed ───────────────
 *
 * The viewer draws everything under a single `Rx-90` scene root plus the
 * "Flip up" toggle (`frame.ts`, §7.3), so a camera read straight off
 * `viewer.camera` is in **viewer** space and would point somewhere else the
 * next time the scene was opened with the flip the other way. A viewpoint is
 * therefore stored in the **dataset** frame — the +Z-up frame spirula wrote,
 * the frame the PLY's `x, y, z` are in, and the frame `cropVolumes.ts` already
 * stores its boxes in — and converted on both sides here.
 *
 * `flip_up` rides along not because the numbers need it but because the *scene*
 * does: restoring a viewpoint saved under the other vertical has to turn the
 * scene back over too, or the camera lands correctly and the world is upside
 * down under it.
 */

export type Vec3 = [number, number, number];

export interface Viewpoint {
  /** Dataset frame. */
  position: Vec3;
  /** Dataset frame — what the orbit controls turn around. */
  target: Vec3;
  /** Dataset frame. The viewer's own up is three.js's +Y, i.e. dataset ±Z. */
  up: Vec3;
  /** Vertical field of view, degrees. */
  fov_y: number;
  /** Which vertical was on when it was saved. */
  flip_up: boolean;
  saved_at: string;
}

/** The scene-root rotation as a quaternion — dataset → viewer. */
function upQuat(flipUp: boolean): THREE.Quaternion {
  return new THREE.Quaternion().setFromAxisAngle(
    new THREE.Vector3(1, 0, 0), upRotationX(flipUp),
  );
}

function toDataset(v: THREE.Vector3, flipUp: boolean): Vec3 {
  const out = v.clone().applyQuaternion(upQuat(flipUp).invert());
  return [out.x, out.y, out.z];
}

function toViewer(v: Vec3, flipUp: boolean): THREE.Vector3 {
  return new THREE.Vector3(...v).applyQuaternion(upQuat(flipUp));
}

/** Read the camera the user has parked, in the dataset frame. */
export function captureViewpoint(
  camera: THREE.PerspectiveCamera, target: THREE.Vector3, flipUp: boolean,
): Viewpoint {
  return {
    position: toDataset(camera.position, flipUp),
    target: toDataset(target, flipUp),
    up: toDataset(camera.up, flipUp),
    fov_y: camera.fov,
    flip_up: flipUp,
    saved_at: new Date().toISOString(),
  };
}

/** The bits of the splat viewer's orbit controls a restore touches. */
export interface OrbitLike {
  target: THREE.Vector3;
  update(): void;
  /**
   * The residual of the last drag. `@mkkellogg/gaussian-splats-3d` builds its
   * OrbitControls with `enableDamping = true` and `dampingFactor = 0.05`, and
   * damped `update()` calls keep applying that residual *after* a teleport —
   * measured, a restore taken 2.5 s after a drag landed 0.025 away from the
   * saved position and drifted on. Its fork exposes these two to clear it;
   * both are optional here so a library that drops them degrades to the
   * approximate restore rather than to a crash.
   */
  clearDampedRotation?(): void;
  clearDampedPan?(): void;
}

/**
 * Put the camera back where it was saved.
 *
 * The near and far planes are left where the framing code set them: they are a
 * property of the scene's size, not of where the camera is standing, and
 * recomputing them from a saved position would undo the sizing that load did.
 */
export function applyViewpoint(
  camera: THREE.PerspectiveCamera,
  controls: OrbitLike | null,
  view: Viewpoint, flipUp: boolean,
): void {
  // Before anything moves: a restore has to land *exactly* on the saved pose,
  // or "Saved" and "Save view" disagree about the same view.
  controls?.clearDampedRotation?.();
  controls?.clearDampedPan?.();
  const target = toViewer(view.target, flipUp);
  camera.position.copy(toViewer(view.position, flipUp));
  camera.up.copy(toViewer(view.up, flipUp).normalize());
  camera.fov = view.fov_y;
  camera.updateProjectionMatrix();
  if (controls) {
    controls.target.copy(target);
    controls.update();
  } else {
    camera.lookAt(target);
  }
}

function triple(raw: unknown, fallback: Vec3): Vec3 {
  if (!Array.isArray(raw) || raw.length !== 3) return fallback;
  if (!raw.every((v) => typeof v === 'number' && Number.isFinite(v))) return fallback;
  return raw as Vec3;
}

/** Rebuild a viewpoint from whatever `settings_json` holds, or null. */
export function fromStored(raw: unknown): Viewpoint | null {
  if (!raw || typeof raw !== 'object') return null;
  const e = raw as Record<string, unknown>;
  const position = triple(e.position, [0, 0, 0]);
  const target = triple(e.target, [0, 0, 0]);
  // The same refusal `core/viewpoint.py` makes: a camera standing on its own
  // target has no direction, and restoring it would leave a blank canvas.
  if (position.every((v, i) => v === target[i])) return null;
  const fov = typeof e.fov_y === 'number' && Number.isFinite(e.fov_y) ? e.fov_y : 50;
  return {
    position,
    target,
    up: triple(e.up, [0, 0, 1]),
    fov_y: Math.min(Math.max(fov, 1), 179),
    flip_up: Boolean(e.flip_up),
    saved_at: typeof e.saved_at === 'string' ? e.saved_at : '',
  };
}

/** One line for the panel, in the dataset frame's own numbers. */
export function describeViewpoint(view: Viewpoint): string {
  const fmt = (v: Vec3) => v.map((c) => c.toFixed(2)).join(', ');
  return `(${fmt(view.position)}) → (${fmt(view.target)}), ${Math.round(view.fov_y)}°`;
}
